.github/workflows/porch-e2e.yml

```diff
@@ -0,0 +1,98 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Porch End-to-End Tests
+on:
+  push:
+    paths-ignore:
+      - "docs/**"
+      - "site/**"
+      - "demos/**"
+      - "Formula/**"
+      - "mdtogo/**"
+      - "package-examples/**"
+      - "release/**"
+      - "scripts/**"
+      - "commands/**"
+      - "internal/cmd**"
+      - "porch/docs/**"
+      - "porch/config/samples/**"
+pull_request:
+    paths-ignore:
+      - "docs/**"
+      - "site/**"
+      - "demos/**"
+      - "Formula/**"
+      - "mdtogo/**"
+      - "package-examples/**"
+      - "release/**"
+      - "scripts/**"
+      - "commands/**"
+      - "internal/cmd**"
+      - "porch/docs/**"
+      - "porch/config/samples/**"
+
+jobs:
+  tests:
+    name: Porch End-to-End Tests
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        image:
+        - "kindest/node:v1.23.4@sha256:0e34f0d0fd448aa2f2819cfd74e99fe5793a6e4938b328f657c8e3f81ee0dfb9"
+
+    steps:
+      - name: Set up Go 1.17
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17
+      - name: Checkout Porch
+        uses: actions/checkout@v2
+      - name: Build kpt
+        run: go install .
+      - name: Build Docker Images
+        run: IMAGE_REPO=porch-kind IMAGE_TAG=${GITHUB_SHA:0:8} make build-images
+        working-directory: ./porch
+      - name: Install KinD
+        uses: engineerd/setup-kind@aa272fe2a7309878ffc2a81c56cfe3ef108ae7d0 # v0.5.0
+        with:
+          version: "v0.12.0"
+          image: ${{ matrix.image }}
+      - name: Load Images
+        run: |
+          kind load docker-image porch-kind/porch:${GITHUB_SHA:0:8}
+          kind load docker-image porch-kind/porch-controllers:${GITHUB_SHA:0:8}
+          kind load docker-image porch-kind/function-runner:${GITHUB_SHA:0:8}
+          kind load docker-image porch-kind/git-server:${GITHUB_SHA:0:8}
+      - name: Install Porch
+        run: |
+          echo ${KUBECONFIG}
+          IMAGE_REPO=porch-kind IMAGE_TAG=${GITHUB_SHA:0:8} make deployment-config
+          kubectl apply --wait --recursive --filename ./.build/deploy
+          kubectl rollout status deployment function-runner --namespace porch-system
+          kubectl rollout status deployment porch-controllers --namespace porch-system
+          kubectl rollout status deployment porch-server --namespace porch-system
+        working-directory: ./porch
+      - name: Wait For Server
+        run: |
+          for i in {1..10}; do
+            sleep $i
+            if kubectl api-resources; then
+              echo "Server is up"
+              break
+            fi
+          done
```

Comment:
Any specific reason for doing the above steps (install kind, install porch, wait) inline here. I was thinking if we can have these steps in a setup script or go-setup-function then it will be easier to run these locally as well right ?

Reply:
Some of the steps seem not applicable to run locally (such as running the engineerd/setup-kind workflow step).

While we can encode some of this in a script (and I'm fine doing it) deep scripts make testing much harder. So for Porch we follow philosophy - environment is set up externally to the test, and then we run go test within the environment.

The e2e test has enough smarts to detect appropriate way to start Git server and will spin up one in-memory or via k8s depending on needs.